### ISSUE TITLE
Update Package Management to Latest Stable NuGet (4.x)

### ISSUE THREAD

**Issue by ilexp** [AdamsLair/duality#574]
### Summary

As a prerequisite of issue #573, a NuGet update will be required. Regardless of this requirement, updating to the latest NuGet for package management is overdue, considering that Duality is still using 2.12, while 4.x has been out for a while.

### Analysis

- Check out the new NuGet client package.
- The package manager classes and tests need to be updated. This will likely be the biggest work item.
  - Address API and behavior changes.
  - Make sure it can consume both Duality packages and regular NuGet library packages.
    - Deal with platform-dependent dependencies and files.
    - Consider writing some more unit tests for this.
  - Consider moving it to a separate project (still part of the Duality editor package in deployment though) to keep direct editor dependencies clean.
- Investigate whether the nightly build needs to be updated.
  - It uses command line nuget, probably the system installed version.
  - Check out which version is used on the CI build and locally on the dev machine.
  - Might be okay as-is, but double-check the used nuget version for publishing packages, as it will soon be required to use the newest one.
- All `.nuspec` files might need to be updated - or they might not. Investigate this.
  - Keep the changes minimal, only do what is required to make and keep them consumable when switching to `netstandard1.1` later on.
  - Create a test project that lets VS 2017 export a `netstandard1.1` nuget package and see what the spec says. Especially the framework paths for files and dependencies should be interesting.
  - Make sure they're still as compatible as possible. Don't use features that aren't required.
  - If changes are required, will need to update dependencies as well, including OpenTK, FarseerDuality, AdamsLair.Winforms, etc.
- Backwards compatibility is key. Don't require a new NuGet for building a package that doesn't need its features.
- Due to the big new set of NuGet-related dependencies, update the `Publish Game` script copy blacklist, as well as potentially existing `.gitignore` files and nightly builder binary package config.
- Consider adding an updated [`.gitignore` file](https://github.com/ilexp/duality-dotgame-sample/blob/ebc95f7cb9c3e2e4b8acbb868fe50793a9712cc2/.gitignore) to the binary release package by default.

**Comment by ilexp** [comment]
Seems like NuGet v3 adopted a micro-dependency model and the former NuGet.Core is no longer updated and locked to v2.x. This could get really ugly - according to the following links, there might be between 24 and 40 dependencies required to get this up an running. 

- https://daveaglick.com/posts/exploring-the-nuget-v3-libraries-part-1
- https://daveaglick.com/posts/exploring-the-nuget-v3-libraries-part-2
- https://daveaglick.com/posts/exploring-the-nuget-v3-libraries-part-3

Investigate further. If it turns out that there will actually be that many different assemblies required in the working directory, consider adding a prerequisite issue to find out if there are ways to keep the main directory clean.

**Comment by ilexp** [comment]
Implementation related to the above blog: https://github.com/Wyamio/Wyam/tree/1897149192a0cb1efe6378aa6185fbaab1eadf9d/src/core/Wyam.Configuration/NuGet

**Comment by ilexp** [comment]
Did a quick test and found that there will be at least 15 individual assemblies required, plus 15 optional XML docs files. This is definitely too much to put into the main directory. Created issue #578 as a dependency for this.

**Comment by ilexp** [comment]
Closed issue #578 after running into more problems than expected during its implementation. Could be re-opened in isolation at some later point, but should no longer be considered a dependency of this issue.

**Comment by ilexp** [comment]
### Immediate ToDo

- Create a separate test project and prototype package discovery, install and potentially updates, should they differ a lot from installs.
  - Set up a test repository of packages, both Duality and others
- Consider moving Duality package management into a separate editor library.
- Update the NuGet dependency to 4.X.
  - Gradually fix all the broken things until it compiles.
  - Gradually fix all the failing tests until it's completely green again.
  - Deal with target platforms and platform-dependent dependencies.
  - Make sure it can consume both Duality packages and regular NuGet library packages.
- Investigate whether the nightly build needs to be updated.
- Investigate whether the `nuspec` files need to be updated, including dependencies outside the main repo.
- Due to the big new set of NuGet-related dependencies, update the `Publish Game` script copy blacklist, as well as potentially existing `.gitignore` files and nightly builder binary package config.
- Consider adding an updated [`.gitignore` file](https://github.com/ilexp/duality-dotgame-sample/blob/ebc95f7cb9c3e2e4b8acbb868fe50793a9712cc2/.gitignore) to the binary release package by default.

**Comment by ilexp** [comment]
### Progress

- Started experimenting with a standalone test project to get a feel for how to use the new NuGet libraries.
  - Metadata discovery works.
  - Simple package install works.
- Created an [issue on the NuGet repository](https://github.com/NuGet/Home/issues/6161) to double-check which of the inconveniences so far are actually by design, which are potentially better in the future, and which are just due to me not understanding NuGet.

### Immediate ToDo

- Create a separate test project to figure out how to use the new NuGet and how to integrate it properly.
  - Consider a custom nuget project implementation.
  - Discover packages by tag.
  - Sandbox NuGet, i.e. not depend on any settings outside the project, especially not package sources.
  - Set up a test repository of packages, both Duality and others.
  - Do some test installs and dependency resolves.
- Consider moving Duality package management into a separate editor library.
- Update the NuGet dependency to 4.X.
  - Gradually fix all the broken things until it compiles.
  - Gradually fix all the failing tests until it's completely green again.
  - Deal with target platforms and platform-dependent dependencies.
  - Make sure it can consume both Duality packages and regular NuGet library packages.
- Investigate whether the nightly build needs to be updated.
- Investigate whether the `nuspec` files need to be updated, including dependencies outside the main repo.
- Due to the big new set of NuGet-related dependencies, update the `Publish Game` script copy blacklist, as well as potentially existing `.gitignore` files and nightly builder binary package config.
- Consider adding an updated [`.gitignore` file](https://github.com/ilexp/duality-dotgame-sample/blob/ebc95f7cb9c3e2e4b8acbb868fe50793a9712cc2/.gitignore) to the binary release package by default.

**Comment by ilexp** [comment]
### Progress

- Cleaned up test project a bit for continued development.
- Continued discussion on the NuGet repo.
- Investigated `ISettings` and `Settings` a bit more.
- Created a custom, readonly settings implementation that does not rely on a config file and emulates the required values. Seems to work so far.

### Immediate ToDo

- Create a separate test project to figure out how to use the new NuGet and how to integrate it properly.
  - Implement a custom NuGetProject.
    - Keep track of installed packages. Maybe just use folder contents?
    - Find the best matching package contents based on the current target framework.
  - Investigate why NuGet queries seem to take so long.
  - Discover packages by tag.
  - Set up a test repository of packages, both Duality and others.
  - Do some test installs and dependency resolves.
- Consider moving Duality package management into a separate editor library.
- Update the NuGet dependency to 4.X.
  - Gradually fix all the broken things until it compiles.
  - Gradually fix all the failing tests until it's completely green again.
  - Deal with target platforms and platform-dependent dependencies.
  - Make sure it can consume both Duality packages and regular NuGet library packages.
- Investigate whether the nightly build needs to be updated.
- Investigate whether the `nuspec` files need to be updated, including dependencies outside the main repo.
- Due to the big new set of NuGet-related dependencies, update the `Publish Game` script copy blacklist, as well as potentially existing `.gitignore` files and nightly builder binary package config.
- Consider adding an updated [`.gitignore` file](https://github.com/ilexp/duality-dotgame-sample/blob/ebc95f7cb9c3e2e4b8acbb868fe50793a9712cc2/.gitignore) to the binary release package by default.

**Comment by ilexp** [comment]
Side note about NuGet packing support for NetStandard packages: http://blog.tdwright.co.uk/2017/11/20/the-perils-of-publishing-a-net-standard-library-to-nuget/

**Comment by ilexp** [comment]
### Progress

- Implemented `GetInstalledPackages` method for custom nuget project based on its installation folder contents.
- Experimented with a first install-update-uninstall test cycle, but didn't get the update part to actually uninstall an old package. Currently inquiring on the [NuGet issue](https://github.com/NuGet/Home/issues/6161) about this.

### Immediate ToDo

- Create a separate test project to figure out how to use the new NuGet and how to integrate it properly.
  - Implement a custom NuGetProject.
    - Find the best matching package contents based on the current target framework.
  - Investigate why NuGet queries seem to take so long.
  - Discover packages by tag.
  - Set up a test repository of packages, both Duality and others.
  - Do some test installs and dependency resolves.
- Consider moving Duality package management into a separate editor library.
- Update the NuGet dependency to 4.X.
  - Gradually fix all the broken things until it compiles.
  - Gradually fix all the failing tests until it's completely green again.
  - Deal with target platforms and platform-dependent dependencies.
  - Make sure it can consume both Duality packages and regular NuGet library packages.
- Investigate whether the nightly build needs to be updated.
- Investigate whether the `nuspec` files need to be updated, including dependencies outside the main repo.
- Due to the big new set of NuGet-related dependencies, update the `Publish Game` script copy blacklist, as well as potentially existing `.gitignore` files and nightly builder binary package config.
- Consider adding an updated [`.gitignore` file](https://github.com/ilexp/duality-dotgame-sample/blob/ebc95f7cb9c3e2e4b8acbb868fe50793a9712cc2/.gitignore) to the binary release package by default.

**Comment by ilexp** [comment]
Putting this up for grabs to indicate that I'd appreciate help on this issue. It's main work will also be nicely self-contained and currently takes place building a prototype in a separate repo, making collaboration a lot easier and less constrained. So if anyone's up for joining in, let me know 👍

**Comment by Barsonax** [comment]
I have a netstandard branch for pathfindax [here](https://github.com/Barsonax/Pathfindax/tree/DotnetStandard). I havent published it to nuget yet though (and probably would have to fix some appveyor related stuff to do so). If you want to test with a real plugin I could publish it to nuget as a prerelease package but as it is now duality does not support prerelease packages. Might be a good moment to add this functionality (I mean this sounds like really simple to add right?).

**Comment by Barsonax** [comment]
Is there any branch I can base my new branch on?

**Comment by ilexp** [comment]
If you're going to look into using NuGet 4.X in Duality, I'd try out the basics in a standalone project first. Check out [this one](https://github.com/ilexp/NuGet4xTest) that I started earlier, it's a regular .Net Framework project. From what I could gather, the main required package is `NuGet.PackageManagement`, which brings in a lot of [required dependencies](https://github.com/ilexp/NuGet4xTest/blob/master/NuGet4XTest/packages.config) automatically. You can then use a `NuGetPackageManager` to preview and execute the [most common operations](https://github.com/ilexp/NuGet4xTest/blob/bf401d18b5ef077145352d72ac1d9751e9404ab1/NuGet4XTest/Program.cs#L125), though I haven't managed to get all of them running 100% yet in a test scenario - updating seems to provide the wrong list of steps for example.

**Comment by ilexp** [comment]
Yep, not sure why exactly. If that's fixed, it should be possible to recreate common operations and use cases in the test project more closely to what the Duality package manager does, and if that works well we might as well start to work on it in a v3 branch.

**Comment by Barsonax** [comment]
Not sure either why it doesn't remove the old package. Maybe you are supposed to do this yourself? Might be worth it to ask them why and how you should fix it before resorting to possible hacks.

Net 4.6 is their next minimum framework I believe. The nuget packages themselves have that info in them so its easy to check.

**Comment by Barsonax** [comment]
Noticed that GetInstalledPackagesInDependencyOrder returns 0 installed packages. Since updating a nuget package also calls this method internally I suspect this is the reason the old package is not removed.

**Comment by ilexp** [comment]
I think you're onto something with `GetInstalledPackagesInDependencyOrder` returning zero. Maybe it has something to do with how the project or repository is defined. Any idea why exactly we get the zero result here?

> EDIT: strangely when I copied all the neccesary code for updating a package of the NuGetPackageManager.cs I linked earlier (of which the commit should match v4.4?) in the test project it did generate a uninstall action as shown below here:

Weird. What happens when you copy the implementation of the [installed packages method](https://github.com/NuGet/NuGet.Client/blob/b6fd263373726dc36680e5035a27deb06a8128ec/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L1191) - does it produce a non-zero result as well?

**Comment by Barsonax** [comment]
Ye when you do that it returns the installed packages correctly. I could give you the modified test code, maybe you will notice something I didn't.

**Comment by Barsonax** [comment]
Solved the package updating issue. The manager was using a different path than the project which caused it to not find the installed packages.

When you change them both to use the ProjectRoot\\Packages folder the uninstall action is generated succesfully. Not sure why its even possible to have a 2 different packages folders in the nugetproject and the packagemanager at the same time when using the API.

**Comment by ilexp** [comment]
> EDIT: implemented a untested solutionmanager and deletemanager but installing, updating and uninstalling a package seems to work now. I have uploaded the code below:

I can add you as a collaborator for the NuGet4X test project so you get push access to the repo. Let me know if that makes sense for you.

> As a side note since from nuget v4.6 on you need .NET 4.6 so maybe its time to start considering updating duality to this version?

Agreed. Let's consider updating editor related projects, but also keep the launcher as-is to ensure a low profile and ability to "just run" on whatever the user has installed.

**Comment by Barsonax** [comment]
I don't think you even want to update the launcher to .net4.6 but to .net core eventually.

If you give me access I will push the necessary changes to the repo.

**Comment by ilexp** [comment]
You should have an invite link 👍

**Comment by Barsonax** [comment]
Found a bug in the CustomNugetProject.GetInstalledPackagesAsync. It seems to wrongly parse the version number for [Pathfindax](https://www.nuget.org/packages/Pathfindax/) since my versioning is slightly different (the last number is the build number). 

![image](https://user-images.githubusercontent.com/19387223/38131713-69a7046a-3408-11e8-9b20-d5b8993e3985.png)

**Comment by ilexp** [comment]
Seems like this is really making progress now, nice work 👍 Roadmap-wise, here's a list of items that we might benefit from looking into before starting to upgrade the package manager itself:

- Check out the basic package operations and verify they work as expected
  - Install packages
  - Uninstall packages
  - Update packages
- Check out the basics of package discovery
  - List all online packages that have a certain tag
  - List all online packages that have a certain ID
  - Retrieve all package metadata for an online package, given ID and version
  - List all locally installed packages
- Investigate why NuGet queries seem to take so long - this was from my early experiments, is this still an issue in the current version?
- Dependency resolve and target frameworks
  - Given ".NET 4.5" / ".NET 4.6" target framework, check if package dependencies resolve correctly and we don't end up with lots of reference assemblies.
  - Given an installed packages metadata, find a way to list package contents that match the current target framework most closely. Find the closest match, without it having to be exact.
- Take a final look at whether there's any way to avoid having to add a 10+ new .dll files to Duality just for this update.

Given that you've already solved some of them, it might not be that far off by now.

> Also I made a new branch 'Barsonax' where I cleaned up some of the nuget code in the Program.cs and moved it into its own class.

Sounds good!

**Comment by Barsonax** [comment]
Added a way to discover duality plugins. I dont see a way to search using just tags as you have to search on the id (you can further filter it client side though). So what iam doing now is finding all packages with 'Duality' in them and then filtering this further on the client.

Performance seems to be good? Not seeing anything strange here.

If you want to avoid adding 10+ dlls to duality just to update nuget we could look into just referencing Nuget.Protocol. This would mean we would have implement alot ourselves though and even with just this package thats still like 9 packages in total.

**Comment by Barsonax** [comment]
> We can't expect all Duality packages to carry Duality in their ID or title. If there's no direct "search by tag", maybe there's a general search that includes tags, which can be used as a first pass, filtering results afterwards?

Maybe there is some domain specific syntax you can use in the searchterm parameter but would have to dig in the nuget source yet again to confirm this. Documentation for the whole nuget lib doesnt seem to exist...

> We need something like this beyond the manual management though, as installed package metadata is retrieved that way. I'm 100% sure NuGet has the info somewhere internally, as it needs to query this when installing, updating or uninstalling, so there should be a way to query this information somehow.

It currently uses the installed packages in the packages folder as package metadata (which is a bit of a hack really). Thats it there is not some other place where this info is kept.

> Good point. I think it's fine if we bump the editor to .NET 4.6 and C# 5.0 when upgrading. Core and launcher are unaffected by this, so we don't lose a lot of compatibility where it's critical.

Consider upgrading to C# 7 while we are at it anyway if we are going to .NET 4.6. I dont think you will lose any compatibity. Any version from VS2012 and up should support this if the correct frameworks and compiler are installed which is in case of VS2017 is by default.

**Comment by Barsonax** [comment]
As I managed to fix all the (local) bugs in the process of porting pathfindax to netstandard I can now push a new (prerelease) version to nuget. Hopefully since iam doing a dual build that targets both .NET4.5 and netstandard2.0 (this is currently the minimum version I can target because of the dependency on duality) it is compatible with the current and future duality packagemanager. Regardless of the outcome it will be good to know how duality is going to handle such a package.

**Comment by ilexp** [comment]
> As I managed to fix all the (local) bugs in the process of porting pathfindax to netstandard I can now push a new (prerelease) version to nuget. Hopefully since iam doing a dual build that targets both .NET4.5 and netstandard2.0 (this is currently the minimum version I can target because of the dependency on duality) it is compatible with the current and future duality packagemanager. Regardless of the outcome it will be good to know how duality is going to handle such a package.
>
> Ofcourse since the current packagemanager does not support prerelease packages the code would have to be modified a bit. I suppose this wouldnt be hard to change just for this test.
> 
> EDIT: pushed it to nuget, now waiting for indexing/validation
EDIT2: the duality packagemanager picks the netstandard version when installing. Everything seems to work though which is good news. Still this makes my dual target build configuration kinda redundant.

Super valuable info, thx for sharing. And pleasantly surprised the .NET Standard package just works with the regular build.

Upgrading the current package manager to distinguish between platform specific content in packages even before upgrading to the new NuGet is on the ToDo for issue #573 as far as I recall, and definitely something to investigate regardless of any other advances.

**Comment by Barsonax** [comment]
Just discovered you can reference duality in a netstandard 1.1 project. See #696.

This is good news is this means you can reference a netstandard library in a pcl and vice versa. So this shouldn't break existing packages.

While this works fine in visual studio it still breaks with the duality package manager. This might very well be because the nuget version is too old.

So gonna check if it works correctly with a newer nuget version. If it works all we have to do is upgrade nuget to support netstandard plugins.

**Comment by Barsonax** [comment]
![image](https://user-images.githubusercontent.com/19387223/47271769-9eb31e80-d57c-11e8-9114-c1bd68f73a54.png)
This is the error you currently get when installing Singularity 0.1.3.68 (which is Netstandard1.1) with the duality package manager.

**Comment by ilexp** [comment]
Hmm, this is actually inside NuGet, probably because the core lib can't handle what's in the package manifest. Can you post a link to your `.nuspec` and `.nupkg` that causes this?

**Comment by Barsonax** [comment]
You can actually download the package from nuget if you go there with the browser.

Thats how I compared the duality packages to others.

Did some work in the weekend to the nuget test repo. Since we will basically be rewriting alot of code we might also consider making DualityEditor less dependend on nuget. I think we should aim to keep the surface area smaller than it is now. Maybe even put a interface and plugin inbetween so there is no coupling anymore.

Also everything in nuget v4 is async and duality code base currently is not. If we adapt it it will spread fast and we need to bump up the C# version. 

Besides why are we still on C#4? Better to switch that to 7.3. When we switch to async and net standard we would lose compatibility with prehistoric VS versions anyway. Better to take advantage of the latest C#.

**Comment by ilexp** [comment]
> Also everything in nuget v4 is async and duality code base currently is not. If we adapt it it will spread fast and we need to bump up the C# version.
>
> Besides why are we still on C#4? Better to switch that to 7.3. When we switch to async and net standard we would lose compatibility with prehistoric VS versions anyway. Better to take advantage of the latest C#.

As far as I know there's no need to bump the version for consuming async methods from a library, only when writing async methods. I'm not a fan of viral code changes where you touch one place and suddenly need to change half the project either, so I'd like to approach this topic in a more "minimal invasive" way, including the use of async code. 

That doesn't mean we can't expand later - but as far as the NuGet update is concerned, it ideally shouldn't change the PackageManager API surface at all and remain transparent to the rest of the codebase. Embracing modern features like async is something that can still happen independently and gradually where it makes sense.

>  I think we should aim to keep the surface area smaller than it is now. Maybe even put a interface and plugin inbetween so there is no coupling anymore.

That layer already exists, it's all the DualityEditor PackageManager code! Besides this, there is no NuGet reference in the codebase.

**Comment by ilexp** [comment]
> We need to await the async methods.

Possibly, but not necessarily - all I'm saying is, it doesn't hurt to look for alternatives first, rather than forcing a codebase-global C# version change because of an implementation detail somewhere. 

`.Wait` is a way to get back to synchronous territory with existing API, but, if choosing to adjust the existing API, it would also be possible to use `ContinueWith` internally where required, and pass the continuation task outside as a return value. There may be other ways, I don't think this has to be a black or white issue.

> I don't think PackageManager.cs should depend on nuget at all. All it needs is something that installs packages.

This sounds like unnecessary abstraction to me. NuGet _is_ something that installs packages, and I don't see we're moving away from NuGet as a delivery mechanism anytime soon - so why put in an extra layer that ends up essentially unused, or worse, even obfuscating what's happening?

An interface / plugin solution would make perfect sense when multiple variants are to be expected, or when system A is actually replaced with system B, to harden the design for future occasions, but neither of this is the case right now. Hiding NuGet as a dependency to user code is totally reasonable as well, but that already happens, so nothing to gain here either.

**Comment by Barsonax** [comment]
> Possibly, but not necessarily - all I'm saying is, it doesn't hurt to look for alternatives first, rather than forcing a codebase-global C# version change because of an implementation detail somewhere.

I don't see the point of not using what the newer C# versions provide. It is a hassle and it will only get worse. Nobody will use VS2010 to contribute to the duality repo. We are not just missing async but alot of other features as well.

> .Wait is a way to get back to synchronous territory with existing API, but, if choosing to adjust the existing API, it would also be possible to use ContinueWith internally where required, and pass the continuation task outside as a return value. There may be other ways, I don't think this has to be a black or white issue.

No just .Wait() will deadlock. Worse this will only happens in a winforms context so unit test won't catch it. You would have to use a different thread to really get around this as this problem is linked to the SynchronizationContext.

> This sounds like unnecessary abstraction to me. NuGet is something that installs packages, and I don't see we're moving away from NuGet as a delivery mechanism anytime soon - so why put in an extra layer that ends up essentially unused, or worse, even obfuscating what's happening?

Its not that I see us moving away from nuget but the current code is very entangled with nuget. It uses nuget types all over the place. Since nuget changed their API so much ALL that code is now broken. If we only had a ~150 lines class implementing purely the stuff we need it would be much easier to fix just that class. We don't control the nuget libraries so its best to keep that surface area as small as possible and not mix duality and nuget concerns. Using a interface could enforce that a bit more (also its easier to test) but is ofcourse not strictly necessary.

**Comment by ilexp** [comment]
> The other task is preventing any nuget types from leaving our abstraction layer and only work with duality types outside of our abstraction layer.

Since it's internal only, I'd make this a "soft goal" only, as in: If it takes too many wrappers or too much code to achieve encapsulation, a leaky abstraction should be enough. It should prevent rewriting big parts of `PackageManager` for a NuGet update, but that doesn't mean every single NuGet type needs to be wrapped. We should keep a reasonable balance.

**Comment by Barsonax** [comment]
Going to try getting the v4 nuget package manager working in duality here:
https://github.com/Barsonax/duality/tree/feature/nuget_package_manager_update_v4

**Comment by Barsonax** [comment]
The new package manager uses a completely different folder structure btw. These are the differences:
- First of all by default it installs to the global nuget package folder. We can change this ofcourse but maybe we shouldn't
- Each targetframework has its own subfolder. There doesn't seem to be a way to go around this. This seems to be tied to how the packages themselves are structured. So there still is a possibility we have to deal with a 'old' style package.

So maybe we should change our approach here. Instead of directly installing the nuget packages to the duality plugins folder I think we should let nuget do its thing and then copy the packages + dependencies to the plugin folder, taking the target framework into account.

The same process happens when you build a csproj. It copies all the needed dll's to the output folder (well unless you turned that off ofcourse).

This does mean we will have to write some code to do this ourselves but that is still probably alot easier than trying to hack around the nuget library. If we use this approach nuget just has to provide us with a path where to find the packages which is much simpler to deal with.

**Comment by ilexp** [comment]
> So maybe we should change our approach here. Instead of directly installing the nuget packages to the duality plugins folder I think we should let nuget do its thing and then copy the packages + dependencies to the plugin folder, taking the target framework into account.

This is already how things work, so we should be in a good starting point here:

1. Duality invokes NuGet API and lets it do its thing - in NuGet 2.x, this meant installing into some folder, which is `Source/Packages` in Duality.
2. Duality uses NuGet API to find out what it installed and where all the files are.
3. Selective copy of those files to their destinations within Duality, i.e. `Plugins`, `Data`, `Source/Code` and the main directory.

What changes is where NuGet puts stuff, and probably how that stuff is structured.

One thing we do have to look out for is how we determine which packages we need to copy or remove: Previously, we could rely on NuGet events for install and uninstall, because they were only called in the context of the local project, not globally. If a global package folder means that those events now work in a global scope, they will no be a reliable indicator on the required steps to install / uninstall / update a package.

**Comment by Barsonax** [comment]
If #710 is merged and if we also update `Nuget.Core` to 2.14 (which is easy as there are no breaking changes from what I have seen) then I dont think we need to update to nuget 4.x.

**Comment by ilexp** [comment]
Some links for picking up this issue again at some point:

- [Samples Repo](https://github.com/NuGet/Samples)
- [NuGet.Client Implementation](https://github.com/NuGet/NuGet.Client/tree/release-4.9.4-rtm/src/NuGet.Core)

No need right now due to our recent progress and NuGet.Core update, just keeping the links around.

**Comment by ilexp** [comment]
Since we've run into trouble with .NET Standard only packages via NuGet 2.14 in issue #573, I'm bumping this issue again with some fresh thought and brief research.

- Reference to [NuGet/8312](https://github.com/NuGet/Home/issues/8312) where we asked to clarify porting and API docs (thx to @Barsonax for bringing it up)
- [NuGet Client SDK](https://docs.microsoft.com/en-us/nuget/reference/nuget-client-sdk) stub page on the Microsoft docs. Nothing much to gain here, but at least we now know what exact term to google for when researching this.
- [Revisiting the NuGet Client Libraries](https://martinbjorkstrom.com/posts/2018-09-19-revisiting-nuget-client-libraries), a blog post going into detail on approaches to use the client SDK in custom projects.

After letting this topic sink in for a while, and also moving forward with NuGet.Core integration in #703, #710 and #715, I think it makes sense to be more open towards lower level implementations here, as @Barsonax suggested before, and ditch the high level `NuGet.PackageManagement` stuff. Our use case is not necessarily always the same as intended by the NuGet / Visual Studio folks, package management is among the best unit-tested areas in Duality, and we've had to do a custom implementation of the 2.14 `PackageManager` anyway to get target frameworks working. Another pro argument would be reducing the number of dependencies and being .NET Standard compatible, as mentioned in the [above blog posting](https://martinbjorkstrom.com/posts/2018-09-19-revisiting-nuget-client-libraries).

**Comment by Barsonax** [comment]
Yup upgrading to Nuget 2.14 only solved some errors but didn't actually made nuget understand how to resolve netstandard properly.

So once more into the breach.

**Comment by Barsonax** [comment]
Started working on a nuget v5.1 implementation for duality (yup thats the latest version currently) based on findings in https://martinbjorkstrom.com/posts/2018-09-19-revisiting-nuget-client-libraries

The good part is that it correctly resolves dependencies as can be seen here:
![image](https://user-images.githubusercontent.com/19387223/61175320-8624e180-a5ad-11e9-9a67-097a93317c1f.png).

The branch can be found here https://github.com/ilexp/NuGet4xTest/tree/Barsonax

Gonna take some work to get all needed functionality though

**Comment by ilexp** [comment]
Awesome 🙂 Already looks better than the previous approach, more commands and work, less interface plumbing and fake project implementations. How are the API docs on all those classes?

**Comment by Barsonax** [comment]
Haven't checked the api docs yet, I took whats in the blog post and made it work for nuget v5. I also added a simple console logging class to help our debugging efforts (I couldn't find one in the library with intelisense but maybe there is one already).

**Comment by Barsonax** [comment]
# Progress
- Added a method to uninstall a package
- Added a method to get all installed packages
- Added a packageconfig class to work with a package config file. Just a simple model class around it with serialization and deserialization functionality.

# TODO
- Install should write to the package config file instead of downloading the packages
- Add a restore method that actually downloads the packages
- Uninstall should also update the package config file
- Add method to search for packages, should support tags
- Add unit tests

**Comment by Barsonax** [comment]
With the above progress I believe we have achieved the basic functionality one would need for implementing a package manager. Next goals:
1. Try to match the public API of this class https://github.com/AdamsLair/duality/blob/master/Source/Editor/DualityEditor/PackageManagement/Internal/NuGetTargetedPackageManager.cs if this is reasonable.
2. Add unit tests to verify stuff is working as expected.
3. Start integrating the code into duality.

@ilexp I already made this project https://github.com/ilexp/NuGet4xTest/blob/Barsonax/NuGet4XTest/NuGet4XTest.csproj a netstandard library. How do you want to integrate this into duality? Just add this project to the duality solution or only the source?

**Comment by ilexp** [comment]
Great work 👍 

> @ilexp I already made this project https://github.com/ilexp/NuGet4xTest/blob/Barsonax/NuGet4XTest/NuGet4XTest.csproj a netstandard library. How do you want to integrate this into duality? Just add this project to the duality solution or only the source?

Just the source - for the time being, I'd prefer integrating the NuGet glue code directly in the internal area of package management.

**Comment by Barsonax** [comment]
> Just the source - for the time being, I'd prefer integrating the NuGet glue code directly in the internal area of package management.

Ok

Another thing. I think nuget now does alot of caching for you. Not sure what duality caches currently but that might be unnecessary.

**Comment by ilexp** [comment]
> Another thing. I think nuget now does alot of caching for you. Not sure what duality caches currently but that might be unnecessary. 

No caching on the Duality side, except for keeping (wrapped) metadata in memory for fast traversal. The `Source/Packages` folder is managed completely by NuGet, so if it doesn't need that anymore, that's fine too. 

Only thing that might be crucial here are events for installed / uninstalled packages which currently only fire for actual changes to the `Source/Packages` folder, so if you install a package that's already there no events are fired. Some of the code may rely on this event behavior, but it could probably be emulated in case this part of the NuGet design is now fundamentally different.

**Comment by Barsonax** [comment]
Started working on integrating it into duality: https://github.com/Barsonax/duality/tree/feature/nugetUpgrade

Running into alot of compiler errors due to nuget references all over the place so its gonna take a while. Also iam not sure how I can fix the PackageCache class as the API it uses no longer exists.

**Comment by ilexp** [comment]
> Started working on integrating it into duality: 

🎉 🙂 

> EDIT: removed PackageCache. Will see if performance degrades or not and then decide to reimplement it or not.

Also watch out for wasteful re-creation of the Duality package info class all the time - might still be worth storing in a private dictionary somewhere. Not having the wrapped info unique _might_ lead to a bug or two as well, not sure if there wasn't any code that relied on reference equality of the info type for the same identity. Could be wrong, but watch out for it.

> EDIT2: I managed to avoid infesting the entire code base with async await by wrapping up some methods with Task.Run which should prevent any deadlocks (using a simple Wait will deadlock in winforms). Once we will switch to async await all the way these can be removed.

Sounds good 👍 I think once this is merged safely, we might just switch to async / await for the isolated case of the package manager and its editor frontend right afterwards. It's not core code, so there's probably less to worry about, and it's a good use case to test the waters separately from all the rest.

**Comment by Barsonax** [comment]
Running into a issue where `UninstallNonRegisteredPackages` keeps uninstalling everything. Seems installed packages are also tracked in PackageSetup class and friends but not sure yet how to make those work with the new nuget logic.

So so far the only thing that works is searching for packages (which btw loads alot faster).

**Comment by ilexp** [comment]
> Seems installed packages are also tracked in PackageSetup class

The Duality package setup is considered the ground truth of what should be installed (deep dependencies implicit), and the package manager will install and uninstall packages in order to match that. This is what triggers package restore, but also allows to just edit the file (or pull a new one from version control for changes) and restart the editor to get a certain setup.

> [...] not sure yet how to make those work with the new nuget logic.

> So so far the only thing that works is searching for packages (which btw loads alot faster). Also had to comment out some code that probably is still needed but needs to be rewritten.

Let me know if you need more info, or someones head to bounce ideas off in any case.

**Comment by Barsonax** [comment]
This is getting a bit messy as I cannot simply use PackageSetup directly because it uses alot of different types for the same nuget concepts:
- LocalPackage -> combines PackageName and PackageInfo
- PackageName -> PackageIdentity
- PackageInfo -> IPackageSearchMetadata
So you end up converting between those all the time. Not pretty.

**Comment by ilexp** [comment]
> This is getting a bit messy as I cannot simply use PackageSetup directly because it uses alot of different types for the same nuget concepts:

Not sure I understand what's wrong yet, please elaborate. 

The Duality package manager intentionally hides away NuGet behind a Duality-only API surface, so both editor and plugins don't depend on NuGet directly, which is part of what makes this update possible in the first place. There are "Duality side" concepts for package management that apply to Duality API for managing packages, and they're expressed through those types. So it's of course expected that some of those Duality types have similarities with NuGet types - but why would that be a problem..?

> Maybe you can take a look at https://github.com/Barsonax/duality/tree/feature/nugetUpgrade and see if iam doing it wrong or we need to clean this up?

Did a quick branch compare, and it seems a bit off, as almost all of the changes seem to do a netstandard switch and project system changes, not NuGet updates. Since we're going to need that for the PR anyway, can you do an isolated branch that only contains the NuGet related changes? Probably easier to talk about too.

**Comment by Barsonax** [comment]
> Not sure I understand what's wrong yet, please elaborate.

I also implemented a packageconfig class myself because I expected that we needed this. However since package setup does this for you I believe I no longer need that class. However I also implemented some logic that uses my class that needs to be refactored/removed now. Basically I made a small design error and got frustrated yesterday 😅

> Did a quick branch compare, and it seems a bit off, as almost all of the changes seem to do a netstandard switch and project system changes, not NuGet updates.

I based it on my netstandard branch as these changes are not compatible with .NET4.5.2. Compare it with that branch should make it alot cleaner.

**Comment by ilexp** [comment]
Ah, got it. We've all been there 😄 Feel free to ping me whenever you need info on existing stuff. My response times admittedly have increased recently, but I'll help where I can 👍 

> Might need to upgrade to .NET 4.7.2 and C# 7.3 (this one should work without build changes?) in a separate branch to avoid a big mega PR.

Sounds good, I'm on it. Will see if I can manage to pull this off today.

**Comment by Barsonax** [comment]
> Sounds good, I'm on it. Will see if I can manage to pull this off today.

Nice, this should make this PR and the netstandard PR a bit smaller. We do have to fix conflicts but that should be doable.

**Comment by ilexp** [comment]
> Might need to upgrade to .NET 4.7.2 and C# 7.3 (this one should work without build changes?) in a separate branch to avoid a big mega PR.

> Sounds good, I'm on it. Will see if I can manage to pull this off today.

Done, merged and released.

Didn't touch your NuGet branch though, maybe check to see if you can re-base (or re-create + cherry pick?) that based on the new master.

**Comment by Barsonax** [comment]
Cherrypicked the nuget changes from the nugetbranch so that branch is now based on based instead of netstandard

**Comment by Barsonax** [comment]
Lately I have been thinking. What if we drop the integrated package manager in the duality editor and simply use the tools that already exist in visual studio:

Pros:
- It would save us a ton of time and pain in maintaining and updating nuget.
- Better supports a command line workflow (for instance users that do not use the editor).
- Free support for multi framework packages on the nuget side.

Cons:
- Not integrated in the editor

A good place for this would be the GameDebugger project. Ofcourse I think GameLauncher would be a better name.

**Comment by Barsonax** [comment]
Closing this as this issue is no longer relevant since we went for the solution in #786
